### PR TITLE
Initial commit for em4x70 support.

### PR DESCRIPTION
**Pull request by cmolson**
Initially, I only have an em4170 variant used for car transponders.
Also known as the ID48 and possibly Magic II.

This first feature is to read the tag id and user memory 1 and 2.

### CODE REVIEW

File: client/src/cmdlfem.c [hunk 1]
```diff
@@ -25,6 +27,7 @@ static command_t CommandTable[] = {
     {"410x",  CmdLFEM410X,  AlwaysAvailable, "EM 4102 commands..."},
     {"4x05",  CmdLFEM4X05,  AlwaysAvailable, "EM 4205 / 4305 / 4369 / 4469 commands..."},
     {"4x50",  CmdLFEM4X50,  AlwaysAvailable, "EM 4350 / 4450 commands..."},
+    {"4x70",  CmdLFEM4X70,  AlwaysAvailable, "EM 4170 /4070 commands..."},
```
Comment (iceman1001) [inline client/src/cmdlfem.c:30]: swap order,  and space

File: client/src/cmdlfem410x.c [hunk 2]
```diff
@@ -10,6 +10,7 @@
 
 #include "cmdlfem410x.h"
 #include "cmdlfem4x50.h"
+#include "cmdlfem4x70.h"
```
Comment (iceman1001) [inline client/src/cmdlfem410x.c:13]: I don't think it belongs here

File: client/src/cmdlfem4x70.c [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+    // envoke reading of a EM4x70 tag which has to be on the antenna because
+    // decoding is done by the device (not on client side)
+
+    bool errors = false;
+    uint8_t cmdp = 0;
+
+    em4x70_data_t etd = {0};
+
+    while (param_getchar(Cmd, cmdp) != 0x00 && !errors) {
```
Comment (iceman1001) [inline client/src/cmdlfem4x70.c:96]: use the cliparser

File: include/em4x70.h [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+//-----------------------------------------------------------------------------
+// Low frequency EM4x70 structs
+//-----------------------------------------------------------------------------
+
+#ifndef EM4X70_H__
+#define EM4X70_H__
+
+typedef struct {
+    bool parity;
+} PACKED em4x70_data_t;
```
Comment (iceman1001) [inline include/em4x70.h:16]: packed struct with a bool??

File: client/src/cmdlfem410x.c (hunk 2 above)
Comment (cmolson) [reply client/src/cmdlfem410x.c]: Right!

File: include/em4x70.h (hunk 4 above)
Comment (cmolson) [reply include/em4x70.h:16]: I sort of copied the em4x50_data_t struct, I removed the items I wasn't using yet. I have added more items for the next pull request that includes writing support.

I'll remove the PACKED attribute in this push.

Comment (iceman1001) [reply include/em4x70.h]: It just didn't make any sense to have an struct with a bool inside.  So its for next PRs  ok.

File: armsrc/em4x70.c [hunk 5]
```diff
@@ -0,0 +1,594 @@
...
+    int num_bytes = len / 8; // We should have a multiple of 8 here
+
+    uint8_t bytes[8];
+
+    for(int i=0;i<num_bytes;i++) {
+        bytes[i] = bits2byte(bits, 8);
+        bits+=8;
+        Dbprintf("Read: %02X", bytes[i]);
+    }
+}*/
```
Comment (iceman1001) [inline armsrc/em4x70.c:417]: remove custom debug code.
there exists already a few  (see appmain.c)

**Review by iceman1001 — CHANGES_REQUESTED** [review]
only convert to cliparser and remove your custom debug stuff and it looks fine

**Comment by iceman1001** [comment]
And you will need to resolve the conflicting file

**Comment by cmolson** [comment]
I will try and add the cliparser change today as well. Thanks!

I named this em4x70 as the transponder/tag I am using is very close to that, with the only difference from what I can see is in the parity of the commands. I will be continuing to update this and have a few more PR's to submit once this large one is in place.

The transponder/tag I am using is referred to ID48 in the car world.

**Comment by iceman1001** [comment]
yeah,  we might need to rename it later on but for now it will do.
You should mention ID48 inside the help text of your `info` command in order to make things clear.

**Comment by cmolson** [comment]
I've updated to use the cliparser. I also added some text about ID48 automotive transponders to the help text.

There was also a small bugfix to the regular EM4x70 devices that require parity bit in the command. We need to drop the most significant bit of the command before adding a parity so that the entire length remains 4 bits.

**Comment by iceman1001** [comment]
As a initial PR,  its looks ok,   most likely some minor things but we sort it out.   There was a spelling mistake in the help text among others.

Thanks for the PR and the new supported type!

**Pull request merged**